### PR TITLE
#166265849 Set up email address verification

### PR DESCRIPTION
**Pull request by encodedBicoding**
### What does this PR do?
1. Using nodemailer as the email sending library, it sets up email notifaction to be sent to a user upon registration
2. Upon registration, It sets the user `email_confirm_code` property to a 64 char random string if the user hasn't verified their email, and to `null` if they have.

#### Description of Task to be completed?
1. Install nodemailer,
2. Configure nodemailer.
3. Creates gmail account with email: `hellobooksnigeria@gmail.com` password:`undefined`.
4.Installs `jsonwebtoken` this enables generation of email unique token for each user
#### How should this be manually tested?
N/A
#### Any background context you want to provide?
I will fail on travis beacuse of the DATABASE_URL environment variable has not been set there.
#### What are the relevant pivotal tracker stories?
[#166265849](https://www.pivotaltracker.com/story/show/166265849)
#### Screenshots (if appropriate)
Screenshot (53)
Screenshot (54)
Screenshot (55)

#### Questions:
N/A

### CODE REVIEW

**Review by chuxmykel — CHANGES_REQUESTED** [review]
Why is the build failing @encodedBicoding?

**Comment by chuxmykel** [comment]
> Why is the build failing @encodedBicoding?

Never mind. I just added the test database

File: package.json [hunk 1]
```diff
@@ -7,7 +7,7 @@
     "start": "node build/index.js",
     "test": "jest --forceExit --verbose",
     "coverage": "jest --forceExit --coverage && cat ./coverage/lcov.info | coveralls",
-    "dev": "cross-env DEBUG=dev && nodemon --exec babel-node server/index.js",
+    "dev": "set DEBUG=dev && nodemon --exec babel-node server/index.js",
```
Comment (chuxmykel) [inline package.json]: Why did you change from cross-env?

File: server/app/helpers/auth.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import jwt from 'jsonwebtoken'
+import { config } from 'dotenv'
+
+config();
+
+const secretKey = process.env.SECRET_KEY;
+
+class Auth {
+	static generateMailToken(payload) {
```
Comment (chuxmykel) [inline server/app/helpers/auth.js:9]: Please check your tab-size. It should be 2 spaces rather than 2 tabs. The tab size is enormous.

File: server/db/models/index.js [hunk 3]
```diff
@@ -2,12 +2,15 @@
 import { readdirSync } from 'fs';
 import { basename as _basename, join } from 'path';
 import Sequelize from 'sequelize';
+import dotenv from 'dotenv';
```
Comment (chuxmykel) [inline server/db/models/index.js]: Why do you need dotenv here?

**Review by chuxmykel — CHANGES_REQUESTED** [review]

File: server/db/models/index.js (hunk 3 above)
Comment (encodedBicoding) [reply server/db/models/index.js]: I felt we called process in that file without including the dotenv, and i don't  know if it would build without it even if it was added through another file

File: package.json (hunk 1 above)
Comment (encodedBicoding) [reply package.json]: It was failing in my end, saying cross-env wasn't recognized

Comment (chuxmykel) [reply package.json]: You should run npm install before using the cross-env. It's a package. Changing it to set will affect those of us using non-windows machines.

Comment (encodedBicoding) [reply package.json]: I actually ran npm install before working, that's why my cors could work. I don't why it didn't work. I would try it again

File: server/db/models/index.js (hunk 3 above)
Comment (chuxmykel) [reply server/db/models/index.js]: This file was generated automatically by sequelize-cli. And we're not exactly using the `.env` in this file. The config happens in the `config.js` file.

File: package.json (hunk 1 above)
Comment (chuxmykel) [reply package.json]: All right bro. Pls do

File: server/db/models/index.js (hunk 3 above)
Comment (encodedBicoding) [reply server/db/models/index.js]: Ok, noted

File: server/app/helpers/auth.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+		return jwt.sign(payload, secretKey, {expiresIn: '365days'})
```
Comment (chuxmykel) [inline server/app/helpers/auth.js]: Your mail says the link will last 1hr but your token is for 365 days?

Comment (encodedBicoding) [reply server/app/helpers/auth.js]: Yeah, was thinking about that, since the user would already have an account and will verify email later, how long do u think before this link expires?

Comment (chuxmykel) [reply server/app/helpers/auth.js]: I think 30 mins is ok sef.

Comment (encodedBicoding) [reply server/app/helpers/auth.js]: Ok, let me just leave it at 1hour

Comment (encodedBicoding) [reply server/app/helpers/auth.js]: I have seen websites do 15days

**Comment by chuxmykel** [comment]
Pls I want to understand what's happening in this feature. Why do you need to insert that code into the database? It's not like we're using it for anything

**Comment by encodedBicoding** [comment]
> Pls I want to understand what's happening in this feature. Why do you need to insert that code into the database? It's not like we're using it for anything

File: server/db/models/index.js (hunk 3 above)
Comment (encodedBicoding) [reply server/db/models/index.js]: it was written in PT board. https://www.pivotaltracker.com/story/show/166265849

File: server/app/helpers/auth.js (hunk 4 above)
Comment (chuxmykel) [reply server/app/helpers/auth.js]: Wow. Really?
All right. 1hr is cool

**Review by chuxmykel — APPROVED** [review]

File: server/app/middlewares/verifyEmail.js [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	/**
+    * @description generates email verification link and send to new user email
+    * @param req express request object
+    * @param res express response object
+    * @returns next() middleware
+    */
+	static async generateEmailLink(req, res, next) {
+		const {firstName, lastName, email, password } = req.body;
+		const token = Auth.generateMailToken({firstName, lastName, email, password});
+		const link = `https://localhost:3000/verifyEmail?token=${token}`;
```
Comment (Hadeneekeh) [inline server/app/middlewares/verifyEmail.js]: Please use the Heroku link provided

File: server/app/middlewares/verifyEmail.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+						   font-weight: bold; 
+						   background-color: #5D45E3`;
+
+		const body = `<section style=${style}>
+						<h2 style='text-align: center; color: #5D45E3 '>Greetings ${firstName}!</h2>
+						<div> 
+							You have successfully registered an account with HelloBooks website.
+							To activate your account, kindly confirm your registration.
+
+							<a href='${link}' style=${linkStyle}> CONFIRM REGISTRATION</a>
```
Comment (Hadeneekeh) [inline server/app/middlewares/verifyEmail.js]: Verify your Email instead of Confirm Registration

**Review by Hadeneekeh — DISMISSED** [review]

File: server/app/middlewares/verifyEmail.js (hunk 5 above)
Comment (chuxmykel) [reply server/app/middlewares/verifyEmail.js]: Seconded

File: server/app/middlewares/verifyEmail.js (hunk 6 above)
Comment (chuxmykel) [reply server/app/middlewares/verifyEmail.js]: I think 'confirm registration' can fly.

File: server/app/middlewares/verifyEmail.js [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+					</section>
+					`;
+
+		const mailObj = {
+			email,
+			subject: 'Verify Email',
+			body,
+		};
+		try {
+			await Mailer.notify(mailObj);
```
Comment (chuxmykel) [inline server/app/middlewares/verifyEmail.js]: Why do you need to await this? I don't see the point. The next line as well, you're chaining a `.then` to the findOne method and also awaiting it?

**Review by chuxmykel — CHANGES_REQUESTED** [review]

File: server/app/middlewares/verifyEmail.js (hunk 5 above)
Comment (encodedBicoding) [reply server/app/middlewares/verifyEmail.js]: where is the heroku link?

File: server/controllers/userController.js [hunk 8]
```diff
@@ -14,31 +16,41 @@ class UserController {
...
+        email_confirm_code: mailToken,
         signupMethod: 'local'
       };
-
       const token = auth.generateToken(user);
-
+    
       const createdUser = await models.Users.create(user);
-      return util.successStatus(res, 201, 'User added successfully', {
+      const link = `https://helobooks-staging.herokuapp.com/api/v1/auth/verifyEmail?token=${mailToken}`
```
Comment (chuxmykel) [inline server/controllers/userController.js]: Hey bro. This is the staging link, Please use the production link 
https://helobooks.herokuapp.com/api/v1/

**Review by chuxmykel — CHANGES_REQUESTED** [review]
Please change the URL to the production URL so we can push the app tonight

**Review by cvjude — APPROVED** [review]

**Review by chuxmykel — APPROVED** [review]
LGTM
